### PR TITLE
Improvements to MacOS builds

### LINKED ISSUE #120: Building for MacOS should encourage setting an SDK version
HID4Java should be built using an older SDK, such as 10.9.

### MacOS x86_64
* At time of writing this, the 10.9 SDK is the oldest SDK which Apple will allow notarization for. 
* Normally the SDK can be [extracted from a XCode 6.2 (apple.com)](http://developer.apple.com/devcenter/download.action?path=/Developer_Tools/Xcode_6.2/Xcode_6.2.dmg).
* Alternately the SDK can be obtained as a stand-alone archive from a mirror such as: https://github.com/phracker/MacOSX-SDKs/releases/download/11.3/MacOSX10.9.sdk.tar.xz (much smaller file size)

### MacOS aarch64
* At the time of writing this, the 11.0 SDK is the oldest SDK which is capable of building for the Apple Silicon/M1 aarch64 CPU.
* Normally the SDK can be [extracted from a XCode 12.2 (apple.com)](https://developer.apple.com/services-account/download?path=/Developer_Tools/Xcode_12.2/Xcode_12.2.xip)
* Alternately the SDK can be obtained as a stand-alone archive from a mirror such as: https://github.com/phracker/MacOSX-SDKs/releases/download/11.3/MacOSX11.0.sdk.tar.xz (much smaller file size)

Configuring the environment must be done in two steps:
* Specifying the full path to the SDK
   ```bash
   # e.g.
   export SDKROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX10.9.sdk 
   ```
* Specifying the target OS
   ```bash
   # e.g.
   export MACOSX_DEPLOYMENT_TARGET=10.9
   ```
* To confirm, you can check the outputted file using `otool -l`
   ```bash
   # e.g.
   otool -l src/main/resources/darwin-x86-64/libhidapi.dylib
   # or (filter relevant portion for 10.9 or 11.0 binaries)
   otool -l src/main/resources/darwin-x86-64/libhidapi.dylib |grep -E 'LC_VERSION_MIN_MACOSX|LC_BUILD_VERSION' -A4
   ```

### LINKED ISSUE #122: build-hidapi.sh: Remove calls to "make clean"
Some targets in `build-hidapi.sh` will fail when `make clean` is called with the error:

```diff
- make: *** No rule to make target `clean'.  Stop.
```

I propose we abolish all calls to `make clean` in `build-hidapi.sh` in favor of reseting the `hidapi` git repository between builds.

Example:

```bash
function git-clean {
  echo -e "Resetting Cpp/hidapi"
  git clean -fd > /dev/null 2>&1 || exit     # remove all untracked files
  git reset --hard > /dev/null 2>&1 || exit  # reset all tracked files
}
```

```diff
- echo -e "${green}Building Windows 64-bit${plain}"
+ echo -e "${green}Building Windows 64-bit${plain}" && git-clean
   dockcross-windows-shared-x64 bash -c 'sudo apt-get update && sudo apt-get --yes install libudev-dev libusb-1.0-0-dev && sudo ./bootstrap && sudo ./configure --host=x86_64-w64-mingw32 && sudo make'
// ...
```

### PR DESCRIPTION
**Pull request by tresf**
Note before merging: This PR contains commits from #115, but they can be refactored out once that branch is merged into `develop`.

Relevant commits:
* 1e2be9a - Add darwin-aarch64 to dockcross
* fba8f73 - Bump JNA to 5.8.0 (allows separate `.dylib` files for separate architectures on MacOS)
* 88c66e9 - Switch to JNA 5.8.0 darwin suffix
* 88c66e9 - Gracefully handle missing folders, files
* 408420c - Remove problematic calls to `make clean` (Closes #122)
* e40bdfe - Add hints for `SDKROOT`, `MACOSX_DEPLOYMENT_TARGET` (Closes #120)

Related #108

### CODE REVIEW

File: build-hidapi.sh [hunk 1]
```diff
@@ -26,316 +37,381 @@
...
+    make CFLAGS="-arch x86_64"
     if [[ "$?" -ne 0 ]]
       then
-        echo -e "\033[31mFailed\033[0m - Removing damaged targets"
+        echo -e "${red}Failed${plain} - Removing damaged targets"
         rm ../../Java/Personal/hid4java/src/main/resources/darwin/libhidapi.dylib
       else
-        echo -e "\033[32mOK\033[0m"
+        echo -e "${green}OK${plain}"
         cp mac/.libs/libhidapi.0.dylib ../../Java/Personal/hid4java/src/main/resources/darwin/libhidapi.dylib
```
Comment (tresf) [inline build-hidapi.sh:325]: Does this need to change to `darwin_x86-64`?

Comment (tresf) [reply build-hidapi.sh:325]: Answering my own question, yes, [it should change](https://github.com/java-native-access/jna/commit/ab2cb98cb5b267d964d465ad95d8161808f18b97), but JNA was courteous to fallback to the old location if missing.  Will patch and resolve this conversation.

**Comment by DaPutzy** [comment]
Hi! I opened [this issue](https://github.com/gary-rowe/hid4java/issues/131) a couple days ago and created a [little patch](https://github.com/gary-rowe/hid4java/files/10047591/added_option_to_open_devices_on_darwin_in_non-exclusive_mode.patch.txt) to add the "non-exclusive device open mode" for mac os.

I feel like this would be a great addition to this PR!

It does require `libhidapi` > `0.12.0` though, and I don't think that is the version currently used in this PR.
I could not figure out where the version is actually coming from though?

**Comment by tresf** [comment]
> Hi! I opened [this issue](https://github.com/gary-rowe/hid4java/issues/131) a couple days ago and created a [little patch](https://github.com/gary-rowe/hid4java/files/10047591/added_option_to_open_devices_on_darwin_in_non-exclusive_mode.patch.txt) to add the "non-exclusive device open mode" for mac os.


Thanks!  Can you please open a separate PR for this patch?

> I feel like this would be a great addition to this PR!

Since it modifies the default behavior I think it should be separate for better tracking, but it definitely improves the quality of life for MacOS users.

I haven't read the PR yet, but many applications will want to maintain exclusive access, so I would recommend (if not already) that the PR has an API toggle for this behavior.

> It does require `libhidapi` > `0.12.0` though, and I don't think that is the version currently used in this PR. I could not figure out where the version is actually coming from though?

AFAIR, the build system always uses the latest version from upstream.

**Comment by DaPutzy** [comment]
I implemented a toggle and the default behaviour is the same as before :)

I build this on top of your PR but i just noticed you only changed the build anyway so there should be no issue creating a separate PR.

**Pull request merged**